### PR TITLE
Switch to Brotli C bindings for CPython, brotlicffi for non-CPython

### PR DESCRIPTION
**Pull request by sethmlarson**
Closes #1620

### CODE REVIEW

File: docs/advanced-usage.rst [hunk 1]
```diff
@@ -334,7 +334,8 @@ Brotli Encoding
 
 Brotli is a compression algorithm created by Google with better compression
 than gzip and deflate and is supported by urllib3 if the
-`brotlipy <https://github.com/python-hyper/brotlipy>`_ package is installed.
+`Brotli <https://pypi.org/Brotli>`_ package or
+`brotlicffi <https://github.com/python-hyper/brotlicffi>`_ package is installed.
```
Comment (pquentin) [inline docs/advanced-usage.rst:338]: Should this read "are installed"? Genuinely not sure as non-native English speaker

**Review by pquentin — CHANGES_REQUESTED** [review]
Looks great! pypy `make_headers` failed because of that change though. I'm not sure what got installed as nox hides the `pip install` log

File: docs/advanced-usage.rst [hunk 2]
```diff
@@ -334,7 +334,8 @@ Brotli Encoding
 
 Brotli is a compression algorithm created by Google with better compression
 than gzip and deflate and is supported by urllib3 if the
-`brotlipy <https://github.com/python-hyper/brotlipy>`_ package is installed.
+`Brotli <https://pypi.org/Brotli>`_ package or
```
Comment (mgorny) [inline docs/advanced-usage.rst:337]: This is 404.

**Comment by mgorny** [comment]
That said, I actually don't like the idea of conditionally depending on two disjoint implementations. This increases the maintenance burden on downstreams, and means that things can go out of sync easily (and therefore become broken on some implementations but not others).

File: docs/advanced-usage.rst (hunk 2 above)
Comment (sethmlarson) [reply docs/advanced-usage.rst:337]: Good catch!

**Comment by sethmlarson** [comment]
@mgory Unfortunately this is the situation we've been in for a long time *implicitly* due to the namespace conflict between brotli and brotlipy. Felt it was better to make explicit instead.

Plus performance wins on CPython and means that CPython users will get the latest and greatest Brotli sooner.

This situation isn't unique to us either, celery and fonttools both do this split requirement and it seems to have worked fine for them. Using only the basic Decompressor API hopefully won't lead to breakages :)

**Comment by mgorny** [comment]
But then, is there any reason to prefer brotlicffi on PyPy? FWICS the Python bindings from brotli build just fine on PyPy3.

I personally would prefer to use brotlicffi unconditionally everywhere because it supports linking against shared libbrotli now which is a major improvement over building multiple statically linked copies of it, and having to rebuild them every time upstream makes a release.

**Comment by sethmlarson** [comment]
@mgorny Good question! I haven't measured the performance differences between `Brotli`/`brotlicffi` on PyPy3.

**Comment by mgorny** [comment]
Anyway, I don't really mind as long as I can use brotlicffi without having to patch the package to support it.

File: docs/advanced-usage.rst (hunk 1 above)
Comment (shazow) [reply docs/advanced-usage.rst:338]: I think the current phrasing works, or could change packages to plural and then "are installed" would work. Either way. :)

**Review by shazow — APPROVED** [review]

**Comment by sethmlarson** [comment]
Thanks all for reviews, I believe I fixed the brotli test on PyPy now :) Will merge once I get all green.

**Review by pquentin — APPROVED** [review]

**Pull request merged**